flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/BackgroundTaskTest.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.adaptive;
+
+import org.apache.flink.core.testutils.FlinkMatchers;
+import org.apache.flink.core.testutils.OneShotLatch;
+import org.apache.flink.testutils.executor.TestExecutorResource;
+import org.apache.flink.util.TestLogger;
+
+import org.hamcrest.Matchers;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+/** Tests for the {@link BackgroundTask}. */
+public class BackgroundTaskTest extends TestLogger {
```

Comment:
We should also add a test that aborting a future that does affect subsequent BackgroundTasks.

Reply:
I don't fully understand which scenario you would like to test. Could you explain it a bit more in detail?

Reply:
Something like `b1 -> b2 -> b3`, blocking on `b1` and then aborting `b2` and checking that after unblocking `b1`, `b3` will be executed?

Reply:
I'm interested in the behavior for the case where we have `b1 -> b2`, and `b1` fails.

I initially read the code such that `b2` would still run, but that is not the case because we use `thenApplyAsync`.

This was a bit surprising actually; the javadocs for the BackGroundTask don't make this clear because they don't contain any notion of failed executions. Conceptually I considered this to just be queue that we pull from with some CompletableFuture sugar.
I'm not sure if the behavior is really desired. Shouldn't background tasks that don't have anything to do with each other be independent? If a dependency is required then the result future should be used to make this explicit?

Reply:
I have extended `testAbortSkipsTasksWhichHaveNotBeenStarted` to cover this case. I hope this does what you envisioned.

Reply:
I think the implementation is as you described it. Tasks are independent of each other and if `b1` fails, then `b2` will be executed. The trick is to use the `termination` future which never transports the failure cause but is only completed once the task has completed.